### PR TITLE
Start the Ably connection when the first trackable is added and stop it when the last trackable is removed

### PR DESCRIPTION
**Pull request by KacperKluka**
Turned off the `autoConnect` option and now starting and stopping of the Ably instance is handled manually. When the first trackable is added (and the first channel is created) the Ably connection is started. When the last trackable is removed (i.e. the last channel is disconnected) the Ably connection is stopped. 
To make sure that those operations are synced and we don't end up in an unexpected state (i.e. having connected channels and closed Ably connection) I've used a mutex approach that checks and modifies the amount of channels in a secure and thread-safe way.

I used the opportunity to change the default `connect()` [a9e93fd](https://github.com/ably/ably-asset-tracking-android/pull/773/commits/a9e93fdc44e24b8224f3e171d3898fa3f0cf5c69) and `disconnect()` [8729e2b](https://github.com/ably/ably-asset-tracking-android/pull/773/commits/8729e2b348a1a4ce93b99241979ac4f5683b4144) methods to suspending ones and change a function to an extension function [aadd1d7](https://github.com/ably/ably-asset-tracking-android/pull/773/commits/aadd1d7b31ac8bb5aab009527998ad16b38073c1).

The main changes (starting and stopping) are included in this commit [3b104fb](https://github.com/ably/ably-asset-tracking-android/pull/773/commits/3b104fb87f9ac3f722f2d872160659b06fea34f5)

### CODE REVIEW

**Review by QuintinWillison — CHANGES_REQUESTED** [review]
Observing the fact that you're having to 'link' your `channels`, `connectingChannelNames` and `channelsMutex` fields together with commentary, I'm left wondering whether it might be better to pull all three of them out into their own class. I imagine:

- them all being `private` fields of that new class
- them having names `channels`, `connectingChannelNames` and `mutex`
- the class having an API specific to the needs of the `Ably` class - e.g.:
  - add a channel to those connecting, taking a channel name and returning whether it's the first channel or not
  - remove a channel from those connecting, taking a channel name and returning whether it was the last channel or not

This will encourage you to only lock the mutex for the minimum amount of time required.

As, for example, I see issues with code readability/maintainability and perhaps durability right now:

- the `stopAblyIfNoChannelsArePresent()` method has to be called with the mutex acquired, but that's not obvious looking at that method's definition in isolation
- you're calling `ably.connectSuspending()` when the mutex is still acquired, which I don't think is necessary

File: common/src/main/java/com/ably/tracking/common/Ably.kt [hunk 1]
```diff
@@ -825,4 +885,33 @@ constructor(
...
+    private suspend fun AblyRealtime.connectSuspending(timeoutInMilliseconds: Long = 10_000L) {
+        if (connection.state.isConnected()) {
+            return
+        }
+        try {
+            withTimeout(timeoutInMilliseconds) {
+                suspendCancellableCoroutine<Unit> { continuation ->
+                    connection.on { channelStateChange ->
+                        when {
+                            channelStateChange.current.isConnected() -> continuation.resume(Unit)
```
Comment (davyskiba) [inline common/src/main/java/com/ably/tracking/common/Ably.kt:916]: Shouldn't we handle an `else` case here?

**Review by davyskiba — CHANGES_REQUESTED** [review]
Agree with Quintin, extracting channel lists and mutex sounds great

File: common/src/main/java/com/ably/tracking/common/Ably.kt (hunk 1 above)
Comment (davyskiba) [reply common/src/main/java/com/ably/tracking/common/Ably.kt:916]: Checked available statuses, and `else` does not sound like a good idea. However are we sure that those are the only two possible outcome states?

**Comment by KacperKluka** [comment]
Closing as we've decided to implement this using our worker queue architecture https://github.com/ably/ably-asset-tracking-android/pull/778

**Pull request closed without merging**